First cut of continuous absorption

alexrlongne opened the pull request:
## Background

* With the cost of particle transport in Artemis, we're looking for ways to reduce the runtime of particles without degrading solution quality. One common method for reducing variance with IMC is to use continuous absorption, where energy weight is continuously absorbed into the material as opposed to sampling distance to absorption events and absorbing all the energy (or a portion of the energy in the implicit scatter method).

* Note this is still a WIP, I still need to think harder about integrating with DDMC 

## Description of Changes

* Add "fraction" to photons field
* Add "fraction" to swarm pack for transport, ddmc and sourcing
* Initialize fraction to 1.0 in sourcing kernel
* Add analog cutoff to transport step
* Attenuate weight exponentially
* Reset fraction to 1.0 at exit of while loop
* Add weight, fraction and absorbed energy to transport and DDMC step args
* Attenuate fraction

## Checklist

- [ ] New features are documented
- [ ] Tests added for bug fixes and new features
- [ ] (@lanl.gov employees) Update copyright on changed files

alexrlongne committed:
First cut of continuous absorption

+ Add "fraction" to photons field
+ Add "fraction" to swarm pack for transport, ddmc and sourcing
+ Initialize fraction to 1.0 in sourcing kernel
+ Add analog cutoff to transport step
+ Attenuate weight exponentially
+ Reset fraction to 1.0 at exit of while loop
+ Add weight, fraction and absorbed energy to transport and DDMC step args
+ Attenuate fraction

RyanWollaeger commented:
This looks great to me so far - guessing DDMC shouldn't be too difficult, since it is the time variable that similarly can be attenuated over. Looks like the variables are wired into the DDMC step function already.

RyanWollaeger commented on src/jaybenne/transport.cpp:
```diff
@@ -225,6 +233,8 @@ TaskStatus TransportPhotons(MeshData<Real> *md, const Real t_start, const Real d
           }
           rng_pool.free_state(rng_gen);
+          // reset fraction of particles that make it census
+          ppack_r(b, ph::fraction(), n) = 1.0;
```
If the fraction gets reset to 1 at census, does it need to be stored as a particle property?

alexrlongne committed:
Merge remote-tracking branch 'origin/develop' into continuous_abs

alexrlongne committed:
Try again to get submodules right

+ Fix error in exp factor
+ Don't use analog in transport
+ Revert changes from DDMC

RyanWollaeger commented on external/parthenon:
Is this submodule update intended?

alexrlongne replied:
Nope! Will fix with a clean checkout and patches since my other attempts have failed.

RyanWollaeger replied:
Thanks for cleaning that up!

RyanWollaeger commented on external/singularity-eos:
Same question - is this intended?

RyanWollaeger commented on src/jaybenne/transport.cpp:
```diff
@@ -181,13 +183,22 @@ TaskStatus TransportPhotons(MeshData<Real> *md, const Real t_start, const Real d
+              Kokkos::atomic_add(&dejbn, e_abs);
+            }
+            else {
+              // process cutoff absorption
```
Would it be worth propagating the particle in analog mode below the threshold, or does that typically not amount to much effect for 1e-6?

alexrlongne replied:
Fixing this PR to process particles below threshold as analog. Full analog can be obtained with an max real cutoff.

RyanWollaeger reviewed (commented):
LGTM - just a few minor questions.

alexrlongne committed:
Fix continuous absorption to be analog below cutoff

+ Add cutoff to jaybenne input parsing
+ Process absorption events and continuous absorption independently

fix compile

alexrlongne committed:
Fix bugs in continuous absroption

+ Tally continuous absorption before processing off-block check
+ Kill zero weight particles before they becomes nans in the comb

alexrlongne committed:
Merge remote-tracking branch 'origin/develop' into continuous_abs

alexrlongne committed:
Fix cout and minor ddmc reordering

alexrlongne committed:
Try checking out hashes at head

RyanWollaeger commented on src/jaybenne/transport_utils.hpp:
```diff
@@ -109,63 +113,117 @@ struct ddmc_step_args {
+
+  // handle face crossings by snapping to face and adding epsilon offset
+  if( event == BOUND) {
+    if (face == 0 || face == 1) {
```
What was the motivation for introducing the face indexing, and removing the dimension checks? (Just curious)

RyanWollaeger commented on src/jaybenne/transport_ddmc.cpp:
```diff
@@ -192,15 +197,33 @@ TaskStatus TransportPhotons_DDMC(MeshData<Real> *md, const Real t_start, const R
+            }
+
+            // continuous absorption with low cutoff allows particles to get to zero energy weights,
+            // kill them so they don't lead to division by zero in population control
```
Oh no - sorry about this - good catch. Would say we can try catching this in the pop. control, but it makes sense to do so here as well.

RyanWollaeger commented on src/jaybenne/transport_ddmc.cpp:
```diff
@@ -192,15 +197,33 @@ TaskStatus TransportPhotons_DDMC(MeshData<Real> *md, const Real t_start, const R
+            // continuous absorption with low cutoff allows particles to get to zero energy weights,
+            // kill them so they don't lead to division by zero in population control
+            if (!(ww > 0.0)) {
+              //std::cout<<"zero weight particle hit killed block"<<std::endl;
```
```suggestion
```
(Test printout?)

RyanWollaeger replied:
Will pull this later - don't want to rerun the CI to remove a comment.

RyanWollaeger commented:
Marked this as ready since the CI may not run with the Draft prefix, according to the README file.

alexrlongne committed:
Apply style to changes
